common/gossip_store.c

```diff
@@ -1,43 +1,119 @@
+	if (!pps->gs)
+		return NULL;
+
+	while (!msg) {
+		beint32_t hdr[2];
+		u32 msglen, checksum;
+		int type;
+
+		if (read(pps->gossip_store_fd, hdr, sizeof(hdr)) != sizeof(hdr)) {
+			per_peer_state_reset_gossip_timer(pps);
+			return NULL;
+		}
 
-	/* FIXME: We should skip over these deleted entries! */
-	msglen = be32_to_cpu(hdr[0]) & ~GOSSIP_STORE_LEN_DELETED_BIT;
-	checksum = be32_to_cpu(hdr[1]);
-	msg = tal_arr(ctx, u8, msglen);
-	if (pread(gossip_store_fd, msg, msglen, offset + sizeof(hdr)) != msglen)
-		status_failed(STATUS_FAIL_INTERNAL_ERROR,
-			      "gossip_store: can't read len %u offset %"PRIu64,
-			      msglen, offset);
+		/* Skip any deleted entries. */
+		if (be32_to_cpu(hdr[0]) & GOSSIP_STORE_LEN_DELETED_BIT) {
+			/* Skip over it. */
+			lseek(pps->gossip_store_fd,
+			      be32_to_cpu(hdr[0]) & ~GOSSIP_STORE_LEN_DELETED_BIT,
+			      SEEK_CUR);
+			continue;
+		}
 
-	if (checksum != crc32c(0, msg, msglen))
-		status_failed(STATUS_FAIL_INTERNAL_ERROR,
-			      "gossip_store: bad checksum offset %"PRIu64": %s",
-			      offset, tal_hex(tmpctx, msg));
+		msglen = be32_to_cpu(hdr[0]);
+		checksum = be32_to_cpu(hdr[1]);
+		msg = tal_arr(ctx, u8, msglen);
+		if (read(pps->gossip_store_fd, msg, msglen) != msglen)
+			status_failed(STATUS_FAIL_INTERNAL_ERROR,
+				      "gossip_store: can't read len %u"
+				      " ~offset %"PRIi64,
+				      msglen,
+				      (s64)lseek(pps->gossip_store_fd,
+						 0, SEEK_CUR));
+
+		if (checksum != crc32c(0, msg, msglen))
+			status_failed(STATUS_FAIL_INTERNAL_ERROR,
+				      "gossip_store: bad checksum offset %"
+				      PRIi64": %s",
+				      (s64)lseek(pps->gossip_store_fd,
+						 0, SEEK_CUR) - msglen,
+				      tal_hex(tmpctx, msg));
+
+		/* Ignore gossipd internal messages. */
+		type = fromwire_peektype(msg);
+		if (type != WIRE_CHANNEL_ANNOUNCEMENT
+		    && type != WIRE_CHANNEL_UPDATE
+		    && type != WIRE_NODE_ANNOUNCEMENT)
+			msg = tal_free(msg);
+	}
 
 	return msg;
 }
 
+/* newfd is at offset 1.  We need to adjust it to similar offset as our
+ * current one. */
+void gossip_store_switch_fd(struct per_peer_state *pps,
+			    int newfd, u64 offset_shorter)
+{
+	u64 cur = lseek(pps->gossip_store_fd, SEEK_CUR, 0);
+
+	/* If we're already at end (common), we know where to go in new one. */
+	if (cur == lseek(pps->gossip_store_fd, SEEK_END, 0)) {
+		status_debug("gossip_store at end, new fd moved to %"PRIu64,
+			     cur - offset_shorter);
+		assert(cur > offset_shorter);
+		lseek(newfd, cur - offset_shorter, SEEK_SET);
+	} else if (cur > offset_shorter) {
+		/* We're part way through.  Worst case, we should move back by
+		 * offset_shorter (that's how much the *end* moved), but in
+		 * practice we'll probably end up retransmitting some stuff */
+		u64 target = cur - offset_shorter;
+		size_t num = 0;
+
+		status_debug("gossip_store new fd moving back %"PRIu64
+			     " to %"PRIu64,
+			     cur, target);
+		cur = 1;
+		while (cur < target) {
```

Comment:
Notice that this can actually skip messages that have not been read yet as well, not just repeat some that have been read. This can in turn lead to incorrect gossip order and fail to sync properly.

Reply:
No, it can't.  

A compaction simply deletes DELETED records, it doesn't reorder them.

The offset_shorter is the amount the final record moved back, by definition, no record can have moved back further than this.

We can end up re-xmitting gossip, but not skipping any.